skorch/net.py

```diff
@@ -602,61 +614,56 @@ def validation_step(self, Xi, yi, **fit_params):
 
         Parameters
         ----------
-        Xi : input data
-          A batch of the input data.
-
-        yi : target data
-          A batch of the target data.
+        batch
+          A single batch returned by the data loader.
 
         **fit_params : dict
           Additional parameters passed to the ``forward`` method of
           the module and to the ``self.train_split`` call.
 
         """
         self.module_.eval()
+        Xi, yi = unpack_data(batch)
         with torch.no_grad():
             y_pred = self.infer(Xi, **fit_params)
             loss = self.get_loss(y_pred, yi, X=Xi, training=False)
         return {
             'loss': loss,
             'y_pred': y_pred,
-            }
+        }
 
-    def train_step_single(self, Xi, yi, **fit_params):
+    def train_step_single(self, batch, **fit_params):
         """Compute y_pred, loss value, and update net's gradients.
 
         The module is set to be in train mode (e.g. dropout is
         applied).
 
         Parameters
         ----------
-        Xi : input data
-          A batch of the input data.
-
-        yi : target data
-          A batch of the target data.
+        batch
+          A single batch returned by the data loader.
 
         **fit_params : dict
           Additional parameters passed to the ``forward`` method of
           the module and to the ``self.train_split`` call.
 
+        Returns
+        -------
+        step : dict
+          A dictionary ``{'loss': loss, 'y_pred': y_pred}``, where the
+          float ``loss`` is the result of the loss function and
+          ``y_pred`` the prediction generated by the PyTorch module.
+
         """
         self.module_.train()
+        Xi, yi = unpack_data(batch)
         y_pred = self.infer(Xi, **fit_params)
         loss = self.get_loss(y_pred, yi, X=Xi, training=True)
         loss.backward()
-
-        self.notify(
-            'on_grad_computed',
-            named_parameters=TeeGenerator(self.module_.named_parameters()),
-            X=Xi,
-            y=yi
-        )
```

Comment:
> in case the user has two or more models (or set of parameters) the process of communicating gradient changes to callbacks is kept transparent.

What would be a use case where the user wants to call `on_grad_computed` explicitly here? E.g. when they only want to clip gradients on a subset of parameters, so that they would only pass `self.module_.SUBMODULE.named_parameters()`? If so, yes, that would require them to override `train_step_single`, which is not nice. I would argue, however, that this filtering should happen on the other side (i.e. in the callback), not here.



> this looks way more standard now

That was the idea: Make this look as much like "normal" PyTorch training code as possible.